baseline-error-prone/src/test/java/com/palantir/baseline/errorprone/OptionalOrElseMethodInvocationTests.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.errorprone;
+
+import com.google.errorprone.BugCheckerRefactoringTestHelper;
+import com.google.errorprone.CompilationTestHelper;
+import org.junit.Before;
+import org.junit.Test;
+
+public final class OptionalOrElseMethodInvocationTests {
+
+    private CompilationTestHelper compilationHelper;
+    private BugCheckerRefactoringTestHelper refactoringTestHelper;
+
+    @Before
+    public void before() {
+        compilationHelper = CompilationTestHelper.newInstance(OptionalOrElseMethodInvocation.class, getClass());
+        refactoringTestHelper = BugCheckerRefactoringTestHelper.newInstance(
+                new OptionalOrElseMethodInvocation(), getClass());
+    }
+
+    @Test
+    public void testNonCompileTimeConstantExpression() {
+        compilationHelper
+                .addSourceLines(
+                        "Test.java",
+                        "import java.util.Optional;",
+                        "class Test {",
+                        "  String f() { return \"hello\"; }",
+                        "  // BUG: Diagnostic contains: non-constant expression",
```

Comment:
what does "non-constant" mean? suspect that's not quite the right word. what you mean here is "method invocation"

Reply:
Well it is, but then as Mark says this doesn't cover operator expressions. Not sure what the right word is?

Reply:
If I understand the `Matchers.contains(ExpressionTree.class, MethodMatchers.anyMethod())` thing correctly, then it errors iff a method is called anywhere inside the `orElse(...)`. A reasonable description would be, for instance, `Expression passed to Optional#orElse must not invoke a method`, or positively (not sure what the right one is), `Expression passed to Optional#orElse invokes a method`.

Reply:
(or `contains a method invocation`, if you prefer). To double-check: `foo.orElse(bar + baz)` is allowed, ja?

Reply:
Well, it is currently but should it be allowed? Are operators ever sufficiently expensive for us to care about? If yes, then we need to think again about our matcher. Perhaps compileTimeConstant OR identifier.

Reply:
String concatenation using operators should be required to execute in a lambda. I don't mind the lambda allocation for math because the optional is already boxing numerical results with similar (avoidable) overhead to creating a lambda.

Reply:
.. then you'll likely see
```
String bar = boom + baz;
return foo.orElse(bar);  // can't concat in here.
```

Reply:
I don't really mind the exact version of the check, as long as you make the text reflect the check. (that's what I was commenting on initially, sorry if this was unclear.)